# Regressão Logística I
## Tarefa II

Vamos trabalhar com a mesma base do exercício anterior, mas vamos aprofundar um pouco mais a nossa regressão.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

import script

import statsmodels.formula.api as smf

In [2]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data'

df = pd.read_csv(url, 
                 names=['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg',
                        'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'])
df['flag_doente'] = (df['num']!=0).astype('int64')
df.head()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3  37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4  41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   

   slope   ca thal  num  flag_doente  
0    3.0  0.0  6.0    0            0  
1    2.0  3.0  3.0    2            1  
2    2.0  2.0  7.0    1            1  
3    3.0  0.0  3.0    0            0  
4    1.0  0.0  3.0    0            0

In [26]:
# Adicionando Flag para Doentes e não Doentes.
df['flag_doente'] = ((df['num'] != 0).astype('int64')
                                     .replace({0: 'Saudavel', 1: 'Doente'}))

# Definindo sex 'Feminino' e 'Masculino' como dado categórico.
df['sex'] = df['sex'].replace({0: 'Feminino', 1: 'Masculino'})

df

age        sex   cp  trestbps   chol  fbs  restecg  thalach  exang  \
0    63.0  Masculino  1.0     145.0  233.0  1.0      2.0    150.0    0.0   
1    67.0  Masculino  4.0     160.0  286.0  0.0      2.0    108.0    1.0   
2    67.0  Masculino  4.0     120.0  229.0  0.0      2.0    129.0    1.0   
3    37.0  Masculino  3.0     130.0  250.0  0.0      0.0    187.0    0.0   
4    41.0   Feminino  2.0     130.0  204.0  0.0      2.0    172.0    0.0   
..    ...        ...  ...       ...    ...  ...      ...      ...    ...   
298  45.0  Masculino  1.0     110.0  264.0  0.0      0.0    132.0    0.0   
299  68.0  Masculino  4.0     144.0  193.0  1.0      0.0    141.0    0.0   
300  57.0  Masculino  4.0     130.0  131.0  0.0      0.0    115.0    1.0   
301  57.0   Feminino  2.0     130.0  236.0  0.0      2.0    174.0    0.0   
302  38.0  Masculino  3.0     138.0  175.0  0.0      0.0    173.0    0.0   

     oldpeak  slope   ca thal  num flag_doente  
0        2.3    3.0  0.0  6.0    0    Saudavel  
1        1.5    2.0  3.0  3.0    2      Doente  
2        2.6    2.0  2.0  7.0    1      Doente  
3        3.5    3.0  0.0  3.0    0    Saudavel  
4        1.4    1.0  0.0  3.0    0    Saudavel  
..       ...    ...  ...  ...  ...         ...  
298      1.2    2.0  0.0  7.0    1      Doente  
299      3.4    2.0  2.0  7.0    2      Doente  
300      1.2    2.0  1.0  7.0    3      Doente  
301      0.0    2.0  1.0  3.0    1      Doente  
302      0.0    1.0    ?  3.0    0    Saudavel  

[303 rows x 15 columns]

In [43]:
tab_sex = script.Qualitativa(df=df, y="flag_doente", x="sex")
tab_sex

flag_doente  Doente  Saudavel  Total  media_doentes  Chance_Odds_Doentes  \
sex                                                                        
Feminino         25        72     97       0.257732             0.347222   
Masculino       114        92    206       0.553398             1.239130   
Total           139       164    303       0.458746             0.847561   

flag_doente  RC_vs_total_OddsRatio    Logito       WOE  
sex                                                     
Feminino                  0.409672 -1.057790 -0.892398  
Masculino                 1.461996  0.214410  0.379802  
Total                     1.000000 -0.165392  0.000000

A descrição das variáveis está recortada abaixo:
- age: idade do paciente em anos
- sex: sexo (1 = male; 0 = female)  
- cp: tipo de dor no peito
  - 1: angina típica
  - 2: angina atípica
  - 3: dor não-angina
  - 4: assintomático
- trestbps: pressão sanguínea em repouso (em mm Hg na admissão ao hospital
- chol: colesterol sérico em mg/dl
- fbs: (açúcar no sangue em jejum > 120 mg/dl) (1 = True; 0 = False)
- restecg: resultados eletrocardiográficos em repouso
  - 0: normal
  - 1: tendo anormalidade da onda ST-T (Inversões de onda T e / ou ST com elevação ou depressão de > 0.05 mV)
  - 2: mostrando hipertrofia ventricular esquerda provável ou definitiva pelos critérios de Estes
- thalach: frequência cardíaca máxima alcançada
- exang: angina induzida por exercício(1 = sim; 0 = não)
- oldpeak = Depressão de ST induzida por exercício em relação ao repouso
- slope: Depressão de ST induzida por exercício em relação ao repouso
  - 1: inclinação ascendente
  - 2: estável
  - 3: inclinação descendente
- ca: número de vasos principais (0-3) coloridos por fluorosopia
- thal: 3 = normal; 6 = defeito corrigido; 7 = defeito reversível
- num: diagnóstico de doença cardíaga (status de doença angiográfica)

In [6]:
df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope         num  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.937294   
std      0.994971   22.875003    0.469794    1.161075    0.616226    1.228536   
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000   
75%      2.000000  166.000000    1.000000    1.600000    2.000000    2.000000   
max      2.000000  202.000000    1.000000    6.200000    3.000000    4.000000   

       flag_doente  
count   303.000000  
mean      0.458746  
std       0.499120  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%       1.000000  
max       1.000000

1. Considere o script que monta a análise bivariada que você fez na tarefa anterior. Transforme esse script em uma função, que deve:
- Ter como parâmetros de entrada:
    - Um *dataframe* contendo os dados a serem avaliados
    - Um *string* contendo o nome da variável resposta
    - Um *string* contendo o nome da variável explicativa
- E deve retornar um *dataframe* com os dados da bivariada. 
**Monte** a mesma bivariada pelo menos três variáveis qualitativas do *data-frame*. Qual delas parece discriminar mais o risco?

2. Monte uma função semelhante para categorizar variáveis quantitativas contínuas (com muitas categorias) como ```age```.  
    Além dos mesmos parâmetros da função anterior, defina mais um parâmetro como número de categorias que você deseja quebrar. Defina um valor '*default*' de 5 grupos para este parâmetro.  

3. Construa um modelo de regressão logística com as variáveis qualitativas: ```sex + cp +  trestbps``` e com a variável quantitativa ```age```.

**Interprete os parâmetros.**

4. Avalie o seu modelo quanto a **calibragem**:
- Calcule a probabilidade de evento predita segundo o seu modelo
- Categorize essa probabilidade em G=5 grupos
- Calcule a probabilidade de evento predita média por grupo
- Calcule a taxa de eventos (média da variável indicadora de eventos) por grupo
- Compare graficamente o valor eperado versus observado para a taxa de maus por grupo

5. Avalie o seu modelo quanto a discriminação calculando acurácia, GINI e KS.

6. tente melhorar o modelo obtido, por exemplo inserindo ou removendo variáveis.  
    Avalie as características do seu modelo (calibragem e acurácia).